util/events/libuv/FileNo.c

```diff
@@ -0,0 +1,258 @@
+/* vim: set expandtab ts=4 sw=4: */
+/*
+ * You may redistribute this program and/or modify it under the terms of
+ * the GNU General Public License as published by the Free Software Foundation,
+ * either version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "util/events/libuv/UvWrapper.h"
+#include "benc/String.h"
+#include "memory/Allocator.h"
+#include "util/events/FileNo.h"
+#include "util/events/libuv/EventBase_pvt.h"
+#include "util/log/Log.h"
+#include "util/Identity.h"
+
+#include <inttypes.h>
+#include <stdio.h>
+
+struct FileNoContext
+{
+    struct FileNo_Promise pub;
+    struct FileNo* fileno;
+
+    Identity
+};
+
+struct FileNo_pvt
+{
+    struct FileNo pub;
+
+    uv_pipe_t server;
+    uv_pipe_t peer;
+
+    /** Job to close the handles when the allocator is freed */
+    struct Allocator_OnFreeJob* closeHandlesOnFree;
+
+    int isActive;
+
+    Identity
+};
+
+static void onClose(uv_handle_t* handle)
+{
+    struct FileNo_pvt* fileno = Identity_check((struct FileNo_pvt*)handle->data);
+    handle->data = NULL;
+    if (fileno->closeHandlesOnFree && !fileno->server.data && !fileno->peer.data) {
+        Allocator_onFreeComplete((struct Allocator_OnFreeJob*) fileno->closeHandlesOnFree);
+    }
+}
+
+#if FileNo_PADDING_AMOUNT < 8
+    #error
+#endif
+#define ALLOC(buff) (((struct Allocator**) &(buff[-(8 + (((uintptr_t)buff) % 8))]))[0])
+
+static void incoming(uv_pipe_t* stream,
+                     ssize_t nread,
+                     const uv_buf_t* buf,
+                     uv_handle_type pending)
+{
+    // Grab out the allocator which was placed there by allocate()
+    struct Allocator* alloc = buf->base ? ALLOC(buf->base) : NULL;
+
+    Assert_true(pending == UV_UNKNOWN_HANDLE);
+
+    if (nread < 0) {
+        uv_close((uv_handle_t*) stream, onClose);
+#ifndef win32
```

Comment:
Because the accepted_fd is only defined in libuv unix stream.c, so (!Defined(win32)) not work at here.